### ISSUE TITLE
R-SAIGE on bioconda

### ISSUE THREAD

**Issue by matuskosut** [weizhouUMICH/SAIGE#272]
Hi @weizhou0 @weizhouUMICH and others,

I managed to revise Saige dependencies and build conda packages for Saige (https://anaconda.org/bioconda/r-saige), such that it could make installation as easier as possible.

### Guide

You will need bioconda and conda-forge channels (added in this order):
```
conda config --add channels bioconda
conda config --add channels conda-forge
```

Feel free to try out the new packages and share feedback:
```
conda create -n saige -c bioconda r-saige=0.44.5
conda activate saige

step1_fitNULLGLMM.R --help
step2_SPAtests.R --help
createSparseGRM.R --help
```

### Available versions

Latest version: ![r-saige](https://anaconda.org/bioconda/r-saige/badges/version.svg) ![latest_date](
https://anaconda.org/bioconda/r-saige/badges/latest_release_date.svg)

- r-saige 0.44.5 (R `r-base=3.6.3`,`r-base=4.0`)

```
conda create -n saige -c conda-forge -c bioconda r-base=4.0 r-saige=0.44.5
conda activate saige
```

### Plaintext dosage files

- r-saige 0.35.8.8 (R `r-base=3.6.3`,`r-base=4.0.3`)

```
conda create -n saige -c conda-forge -c bioconda r-base=4.0 r-saige=0.35.8.8
conda activate saige
```

### Other versions previously released

- r-saige 0.39 (R `r-base=3.5.1`, `r-base=3.6.1`,`r-base=4.0`)
- r-saige 0.42 (R `r-base=3.6.1`,`r-base=4.0`)
- r-saige 0.42.1 (R `r-base=3.6.1`,`r-base=4.0`)
- r-saige 0.43.0 (R `r-base=3.6.1`,`r-base=4.0`)
- r-saige 0.44.0 (R `r-base=3.6.3`,`r-base=4.0`)
- r-saige 0.44.1 (R `r-base=3.6.3`,`r-base=4.0`)
- r-saige 0.44.2 (R `r-base=3.6.3`,`r-base=4.0`)
- r-saige 0.44.5 (R `r-base=3.6.3`,`r-base=4.0`)

To create conda environment with specific R version:
```
conda create -n saige -c conda-forge -c bioconda r-base=3.6.1 r-saige=0.43.0 savvy=1.3.0
conda activate saige
```

Versions 0.44.0 or older might require `savvy=1.3.0`. You will need to install them as shown in the example above, if you experience this issue:
```
... libsavvy.so: cannot open shared object file: No such file or directory
Execution halted
```

For those installing r-saige with R 3.6.1 I recommend installing lower version of optparse (this doesn't apply for R4.0):
```
conda install -c conda-forge "r-optparse=1.6.2=r36h6115d3f_1"
```

### Related

Some of you may also be interested in other packages that were added:
- https://anaconda.org/bioconda/regenie
- https://anaconda.org/conda-forge/bgenix
- https://anaconda.org/bioconda/savvy
- https://anaconda.org/conda-forge/r-metaskat
- https://anaconda.org/conda-forge/cget

**Comment by hannahpky** [comment]
Hi @matuskosut thank you very much for sharing these conda packages for saige. I've been attempting to run  step1 of saige using the step1_fitNULLGLMM.R with optparse R command line parser package; however, I keep receiving segfault errors: " *** caught segfault *** address 0x19, cause 'memory not mapped' " ([saige_error.txt](https://github.com/weizhouUMICH/SAIGE/files/5548083/saige_error.txt)
). This error makes me wonder if it's something to do with my install configuration (i.e. saige is unable to access a necessary memory location). I followed the github instructions to create a conda environment (https://github.com/weizhouUMICH/SAIGE/wiki/Genetic-association-tests-using-SAIGE#installing-saige) and then used your conda r-saige=0.42.1 version with r-base=3.6.1. Any ideas on how to troubleshoot? I'm a newbie when it comes to these errors, so any help would be appreciated. Thank you!

**Comment by matuskosut** [comment]
Hi @hannahpky, thanks for the feedback. Based on the saige_error.txt that you shared it looks like an incompatibility between optparse and r-base version. Could you shared the exact versions of r-optparse and r-base? There is a special hash string that identifies the build number of that version, for example: "r-base 3.6.1 h8900bf8_2". You can get them using these commands in your conda environment:
```
conda list | grep optparse
conda list | grep r-base
```

I think that if we manage to pinpoint the matching version of optparse that we could make it work and perhaps also set a requirement in the package such that it could work for others.

**Comment by matuskosut** [comment]
@hannahpky could you maybe try to install lower version of optparse? e.g.:
```
conda install -c conda-forge "r-optparse=1.6.2=r36h6115d3f_1"
```

**Comment by hannahpky** [comment]
Thanks Matúš - I installed a the lower version of optparse (1.6.2). I no longer receive the following message; however, I am still receiving the same " *** caught segfault *** address 0x19, cause 'memory not mapped'" message.
```
Loading required package: optparse 
Warning message:
package ‘optparse’ was built under R version 3.6.3
```
I included my conda info and list details to help troubleshoot ([conda_info.txt (https://github.com/weizhouUMICH/SAIGE/files/5549836/conda_info.txt)[conda_list.txt](https://github.com/weizhouUMICH/SAIGE/files/5549837/conda_list.txt)). I'm wondering if I need a different r-base version or some other package version?

**Comment by matuskosut** [comment]
@hannahpky interesting, I would definitely try r-base=4.0 or check other version of saige, e.g. 0.42 is available. I will try to have a closer look on this, any other info about inputs (types, parameters,..) that you could share?

**Comment by hannahpky** [comment]
I tried r-base=4.0 with r-saige=0.42.1, but received the same error. These are the inputs I've been using: 
```step1_fitNULLGLMM.R \
--plinkFile=/vgipiper05/hannah/stuttering/population_analysis/merged/potential_ctrls/selected_controls_taketwo_09092020/ALL_stut_merged_casesmatchedctrls_LDprunednoHLAinv \
--phenoFile=/vgipiper05/hannah/stuttering/population_analysis/merged/potential_ctrls/selected_controls_taketwo_09092020/PHENO.txt \
--phenoCol=STUT \
--covarColList=PC1,PC2,PC3,PC4,PC5,PC6 \
--sampleIDColinphenoFile=IID \
--traitType=binary \
--nThreads=4 \
--LOCO=FALSE \
--outputPrefix=/vgipiper05/hannah/stuttering/population_analysis/merged/potential_ctrls/selected_controls_taketwo_09092020/SAIGE_11162020/Stut_SAIGE_fitNullGLMM```

All the inputs appear to be read-in correctly and then immediately afterward I receive the "*** caught segfault ***
address 0x19, cause 'memory not mapped' " error.

**Comment by matuskosut** [comment]
@hannahpky I have got the new version in, could you try to upgrade or maybe recreate your conda environment such that new version is installed..

**Comment by Esther19960123** [comment]
> createSparseGRM.R --help

It works well so I installed saige successfully which I spent two days installing and failed. Thank you a lot

**Comment by hannahpky** [comment]
Yep, thanks @matuskosut the new r-saige version seems to be working well! I'm no longer receiving segfault errors :)

**Comment by weizhou0** [comment]
Hi @matuskosut, 

Thank you so much for sharing the bioconda SAIGE!!!

Wei

**Comment by alecmchiu** [comment]
@matuskosut 

Works like a charm! Spent several days trying to install this on the cluster that I work on to no avail. However, your bioconda installation was able to get me to all the help messages successfully! Hopefully it runs without a hitch too!

**Comment by matuskosut** [comment]
@alecmchiu awesome! You probably installed `0.42.1` which now seems pretty stable. But as I write this a new release `0.43.0` is being uploaded in case you are interested.

**Comment by huanghe0220** [comment]
Hi @matuskosut ,I have followed your steps to install saige.
`conda create -n saige -c conda-forge -c bioconda r-base=3.6.1 r-saige=0.43.0` 
`conda install -c conda-forge "r-optparse=1.6.2=r36h6115d3f_1"`
However, when I came to this step: 
`step1_fitNULLGLMM.R --help`
I received a bug:
`Error: package or namespace load failed for ‘SAIGE’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/home/huanghe/miniconda3/lib/R/library/SAIGE/libs/SAIGE.so':
  libsavvy.so: cannot open shared object file: No such file or directory
Execution halted`

Do you know how to solve this problem? I have been stuck with this package for 1 week, so any help would be appreciated. Thank you!

**Comment by huanghe0220** [comment]
@matuskosut Thanks for your reply. I suppose the saige has been installed successfully. But when I came to step 2 with the example dataset, `Rscript step2_SPAtests.R --dosageFile=./input/dosage_10markers.txt --dosageFileNrowSkip=1 --dosageFileNcolSkip=6 --dosageFilecolnamesSkip=CHR,SNP,CM,POS,EFFECT_ALLELE,ALT_ALLELE --minMAF=0.0001 --sampleFile=./input/sampleIDindosage.txt --GMMATmodelFile=./output/example.rda --varianceRatioFile=./output/example.varianceRatio.txt --SAIGEOutputFile=./output/example.plainDosage.SAIGE.txt --numLinesOutput=2 --IsOutputAFinCaseCtrl=TRUE`
I came into this bug:
`Error in getopt_options(object, args) :
  Error in getopt(spec = spec, opt = args) :
  long flag "dosageFile" is invalid`
If I use my own data, I would bump into another bug in step 1(`Rscript step1_fitNULLGLMM.R`):
`Error in terms.formula(formula, data = data) :
  invalid model formula in ExtractVars`
Do you know why? Looking forward to your reply. Thanks!

Best wishes,
Huang He

**Comment by matuskosut** [comment]
@huanghe0220 is this plain text dosage files being used? If so, you need to use this version `r-saige=0.38.8.8`.

**Comment by huanghe0220** [comment]
@matuskosut I think you mean `r-saige=0.35.8.8`? I couldn't find 0.38.8.8, and installed version 0.35.8.8, but I still got the bug in step 2 with example dataset, and the same bug in step 1 with my own dataset. I got really confused :(

**Comment by matuskosut** [comment]
@huanghe0220 yes, thanks for correcting. I fixed the version typo in the issue text.

I see that your step 2 issue was `long flag "dosageFile" is invalid`  and looking at the source code of 0.35.8.8 `--dosageFile` is present in step2: https://github.com/weizhouUMICH/SAIGE/blob/0.35.8.8/extdata/step2_SPAtests.R (it is not present in newer versions)

Could you make an issue for this if you will still have problem? https://github.com/weizhouUMICH/SAIGE/issues/new

**Comment by huanghe0220** [comment]
@matuskosut Thanks for your reply. I will make an issue.

**Comment by weizhouUMICH** [comment]
Hi @matuskosut, 

There seems the query issue for bgen file in the current bioconda version. I wonder whether you used the bgenix version in the /thirdparty folder when you compile SAIGE. If not, that would cause the issue. Thanks!

Wei

**Comment by matuskosut** [comment]
Hi @weizhouUMICH,

Is this related to issue: #393 ? Were you also able to reproduce? Could it be possible to reproduce with some nonsensitive data? since I don't have access to any datasets.

I built bgenix based on same version as in /thirdparty. I was applying the patches that I identified in your repo. I noticed only one that was missing - although I think it is unrelated - but still I rebuilt bgenix and saige and published in conda in case you want to try it. 

It is possible to also try bgenix separately to isolate the issue: https://anaconda.org/conda-forge/bgenix

**Comment by weizhou0** [comment]
Hi @matuskosut <https://github.com/matuskosut>,

Yes this is related to issue: #393
<https://github.com/weizhouUMICH/SAIGE/issues/393> ? and another user also
observed the same issue using the bioconda version.
The code of bgen in the third party actually is not the same as the bgenix
library https://anaconda.org/conda-forge/bgenix, because I made some
modifications for querying with variant IDs in the gene-based tests.
If SAIGE was compiled using the bgen code in the thirdparty folder, it
should not have the issue. I just updated the version to 0.45. Would you
mind updating the bioconda version?
Thanks so much for your help! I'm also trying to learn how to build the
bioconda library. :)

Thanks,
Wei

**Comment by matuskosut** [comment]
Hi @weizhouUMICH (@weizhou0 ),

One thing to explain is the principle from conda-forge and bioconda - the package should contain only one tool/library which is also implied by the rule that there should not be multiple licenses packaged together. So each package has to be built separately and stated clearly as dependency. Hence we have to build bgenix from separate repo. 
Maybe you could have a look? This is the branch for 1.1.4 version that is used for R-SAIGE: 

https://github.com/huntdatacenter/rbgen/tree/build-v1.1.4

Let me know in case you spot any difference. I was aware of some of your earlier changes and passed the patches into that branch, plus I was also patching waf to run with python3. Since this problem appeared only now, I think it still might be some other cause.

In the meantime conda package for 0.45 was released, check here: https://anaconda.org/bioconda/r-saige

**Comment by weizhouUMICH** [comment]
Hi @matuskosut, 

We have just released a new version 1.0.0. It no longer depends on the bgen library. 

It has computational efficiency improvements for both Step 1 and Step 2 for single-variant and set-based tests. We have created a new program github page https://github.com/saigegit/SAIGE with the documentation provided https://saigegit.github.io/SAIGE-doc/

Thanks!
Wei
